### DISCUSSION TITLE
Generate types for Params on compile

### DISCUSSION THREAD

**Discussion by mikeybinns** [remix-run/remix#5096] [Proposals]
Currently, the type for the Params type is very loose, just "string | undefined", but in most cases, this isn't accurate.

If Remix is already converting the filenames into a route object, it should be possible to calculate the Params value for any page.

Take this example folder structure:
```
> page.$param1.tsx
> page.tsx
```
For this example, we know `$param1` must be a string in `page.$param1.tsx` because if it's undefined, Remix will match `page.tsx` instead, however currently, we need to add a `if (params.param1) {}` in the actions / loaders on the `page.$param1.tsx` page because of the expected type. 
In fact, I'm pretty sure that if page.tsx isn't defined, it would still take you to a 404, not `page.$param1.tsx` because it's not an optional parameter.

Remix likely doesn't know the route structure until the site is being compiled with remix build, so I suspect if we can generate this type, it would need to be done as part of the compiler, unless there's some fancy TS syntax I'm not aware of.

There's precedent for this kind of type compiling, it's what `@prisma/client` uses to generate types from your schema.
Correction: Prisma generates the types when you run its npm commands, so perhaps this can be a remix command instead?

**Comment by aniravi24** [comment]
Currently this is being solved by another library: https://github.com/sandulat/routes-gen

You can make a light wrapper around most or all of the routing using the types generated from this library.

But it would definitely be awesome to figure out how to get even more type safety built into the main project itself. The more type safety the better!

**Reply by mikeybinns** [reply]
Awesome, I didn't know about this library.

It's a shame we have to use `params as Type` because that can lead to incorrect types being applied.

I think I'm going to make a wrapper type for ActionArgs and LoaderArgs for now, something like this:
```ts
// user.d.ts

type ActionArgsWithParams<RouteId extends string> = Omit<ActionArgs, "params"> & {
  params: RouteParams[RouteId],
}
type LoaderArgsWithParams<RouteId extends string> = Omit<LoaderArgs, "params"> & {
  params: RouteParams[RouteId],
}
// routes/products/$productId.tsx
export function action({ request, params }: ActionArgsWithParams<"/products/:productId">) {}
export function loader({ request, params }: LoaderArgsWithParams<"/products/:productId">) {}
```

There would be a lot of value if this was more strictly typed in Remix.

**Reply by machour** [reply]
Side note, you can also use `DataFunctionArgs` instead of LoaderArgs/ActionsArgs in remix

**Reply by mikeybinns** [reply]
While that's accurate, I think I'll stick with the individual ones for now, just in case an update down the road makes LoaderArgs and ActionArgs different from eachother :)

**Reply by aniravi24** [reply]
@mikeybinns yup that's exactly what I did too!
https://github.com/sandulat/routes-gen/discussions/25#discussioncomment-4450787

**Comment by runoncedev** [comment]
I was reading the [param type safety section on the docs](https://remix.run/docs/en/1.16.1/guides/data-loading#param-type-safety):

> Because these params come from the URL and not your source code, you can't know for sure if they will be defined. That's why the types on the param's keys are string | undefined.

And I couldn't understand _how_ could a non-optional param be `undefined`.

After searching a little I ended up in this discussion and I totally agree with OP. It would be awesome if Remix could type this params as just `string` so that we don't need to use `invariant` or any other workaround.

**Reply by kiliman** [reply]
In my experience, params aren't always strings. They could be numbers, for example. So you'd have to do type conversion anyway. That's why I created `remix-params-helper`.

```ts
// routes/users/$userId.tsx
export function loader({ params }: LoaderArgs) {
  const { userId } = getParamsOrThrow(params, z.object({ userId: z.number() })
        //^? userId: number
}
```

**Reply by mikeybinns** [reply]
> In my experience, params aren't always strings. They could be numbers, for example. So you'd have to do type conversion anyway. 

Wouldn't it still return a string you then have to convert to a number, same as how URL search params are always strings too?

**Reply by kiliman** [reply]
My helper does the type conversion, so the app will only see `number`. If the param is invalid, it will `throw new Response('Bad Request', { status: 400 })`

Relying on Typescript types for user input is error-prone. The only way to ensure you have valid input is to use runtime type checking like `invariant` or `zod`.

**Reply by kiliman** [reply]
> ... same as how URL search params are always strings too?

`remix-params-helper` works with `URLSearchParams`, `FormData`, and params (or any object type).

**Reply by machour** [reply]
Haven't had the occasion to try @kiliman's lib yet.
However, I'm using https://github.com/sandulat/routes-gen and this little helper in my `remix.env.d.ts` file:

```ts
import type { DataFunctionArgs } from "@remix-run/node";
import { RouteParams } from "routes-gen";

export type DataFunctionWithParams<RouteId extends keyof RouteParams> = Omit<
  DataFunctionArgs,
  "params"
> & {
  params: RouteParams[RouteId];
};
```

to type my loader/action args

![Screenshot 2023-05-30 at 3 13 56 PM](https://github.com/remix-run/remix/assets/304450/58460ec1-b757-4238-8be6-168c3f97d636)

Works like a charm 👌🏼

**Reply by mikeybinns** [reply]
I've used routes-gen as mentioned above in the thread, but I think we shouldn't need it to be a separate package if Remix's compiler already knows the route map. Is it not possible for Remix itself to generate a more accurate type based on the file name information provided? I understand this would only work after a compile or after running a command to generate the param types, but this seems like a useful feature for Remix to have baked in.

It would still only return "string", so you'd have to do any conversions from a string to, for example, a number yourself, but in most cases, all you need is a string, and it can only ever return a string, it can't return undefined, so the type that Remix returns is technically incorrect, unless I'm missing something here?

As @oncet says "I couldn't understand how could a non-optional param be undefined." and I can't either, it shouldn't be possible, because even if the user manipulates the URL, that would change which file it would load.

Again, if I'm misunderstanding the complexity here, please do let me know, but considering Remix is already decifering the route map, it doesn't seem like that much of a stretch to then output a params type, to be consumed in a similar way to routes-gen. You could even add an optional generic so it wouldn't be a breaking change, not providing a route to the type would return the current type.

**Reply by kiliman** [reply]
Technically you can access params from any route, so a parent route can still access params from child routes (perhaps you have some conditional logic). That's why params are typed as `string | undefined`.
